programs/storage_api/src/storage_processor.rs

```diff
@@ -45,120 +118,200 @@ pub fn process_instruction(
+                    // keyed_accounts[1] should be the main storage key
+                    // to access its data
+                    Err(InstructionError::InvalidArgument)?;
                 }
 
-                let segment_index = get_segment_from_entry(entry_height);
-                if storage_contract.previous_proofs[segment_index].len() != proof_mask.len() {
+                if let StorageContract::ValidatorStorage {
+                    reward_validations, ..
+                } = &mut storage_contract
+                {
+                    let claims_index = get_segment_from_entry(entry_height);
+                    let _num_validations = count_valid_proofs(&reward_validations[claims_index]);
+                    // TODO can't just create lamports out of thin air
+                    // keyed_accounts[0].account.lamports += TOTAL_VALIDATOR_REWARDS * num_validations;
+                    reward_validations.clear();
+                } else if let StorageContract::ReplicatorStorage {
+                    reward_validations, ..
+                } = &mut storage_contract
+                {
+                    // if current tick height is a full segment away? then allow reward collection
+                    // storage needs to move to tick heights too, until then this makes little sense
+                    let current_index = get_segment_from_entry(tick_height);
+                    let claims_index = get_segment_from_entry(entry_height);
+                    if current_index <= claims_index || claims_index >= reward_validations.len() {
+                        println!(
+                            "current {:?}, claim {:?}, rewards {:?}",
+                            current_index,
+                            claims_index,
+                            reward_validations.len()
+                        );
+                        return Err(InstructionError::InvalidArgument);
+                    }
+                    let _num_validations = count_valid_proofs(&reward_validations[claims_index]);
+                    // TODO can't just create lamports out of thin air
+                    // keyed_accounts[0].account.lamports += num_validations
+                    //     * TOTAL_REPLICATOR_REWARDS
+                    //     * (num_validations / reward_validations[claims_index].len() as u64);
+                    reward_validations.clear();
+                } else {
                     return Err(InstructionError::InvalidArgument);
                 }
+            }
+            StorageInstruction::ProofValidation {
+                entry_height: proof_entry_height,
+                proofs,
+            } => {
+                if let StorageContract::ValidatorStorage {
+                    entry_height: current_entry_height,
+                    lockout_validations,
+                    ..
+                } = &mut storage_contract
+                {
+                    if keyed_accounts.len() == 1 {
```

Comment:
<= to be extra safe? maybe it's not needed.

Reply:
If there's no replicator account passed in there's not point running this code :D Basically no proofs I can check this validator's proofs_verification against

Reply:
ok, that's fine.

Reply:
I mean checking keyed_acounts.len <= 1, but keyed_accounts.len can probably not be 0 here.